File: test/buyback_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flipradar/buyback.dart';

BuybackOffer offer({
  required double price,
  BuybackCondition condition = BuybackCondition.likeNew,
  double confidence = 0.98,
  bool uncertain = false,
  DateTime? checkedAt,
  Uri? offerUrl,
}) {
  return BuybackOffer(
    providerId: 'provider-$price',
    providerName: 'Provider',
    productId: 'iphone-15-pro-256',
    matchedTitle: 'Apple iPhone 15 Pro 256 GB',
    condition: condition,
    price: price,
    currency: 'EUR',
    offerUrl: offerUrl ?? Uri.parse('https://example.com/offer'),
    checkedAt: checkedAt ?? DateTime.parse('2026-09-16T08:30:00+02:00'),
    requiresInspection: true,
    matchConfidence: confidence,
    conditionUncertain: uncertain,
  );
}

Map<String, dynamic> payload() => {
      'provider_id': 'example',
      'provider_name': 'Example',
      'product_id': 'iphone-15-pro-256',
      'matched_title': 'Apple iPhone 15 Pro 256 GB',
      'condition': 'like_new',
      'price': 620,
      'currency': 'EUR',
      'offer_url': 'https://example.com/offer',
      'checked_at': '2026-09-16T08:30:00+02:00',
      'price_kind': 'indicative_buyback',
      'requires_inspection': true,
      'match_confidence': 0.98,
    };

void main() {
  test('condition wire values round-trip', () {
    for (final condition in BuybackCondition.values) {
      expect(BuybackConditionWire.tryParse(condition.wireValue), condition);
    }
  });

  test('selects highest eligible offer for the requested condition', () {
    final best = bestComparableBuybackOffer(
      [offer(price: 590), offer(price: 625), offer(price: 610)],
      condition: BuybackCondition.likeNew,
    );
    expect(best?.price, 625);
  });

  test('does not compare different conditions', () {
    final best = bestComparableBuybackOffer(
      [offer(price: 600), offer(price: 700, condition: BuybackCondition.usedGood)],
      condition: BuybackCondition.likeNew,
    );
    expect(best?.price, 600);
  });

  test('excludes uncertain and low-confidence matches', () {
    final best = bestComparableBuybackOffer(
      [offer(price: 700, uncertain: true), offer(price: 680, confidence: 0.75), offer(price: 620)],
      condition: BuybackCondition.likeNew,
    );
    expect(best?.price, 620);
  });

  test('excludes zero-value offers from comparison', () {
    final best = bestComparableBuybackOffer(
      [offer(price: 0), offer(price: 620)],
      condition: BuybackCondition.likeNew,
    );
    expect(best?.price, 620);
    expect(offer(price: 0).isEligibleForComparison, isFalse);
  });

  test('excludes implausible prices and unsafe offer URLs', () {
    expect(offer(price: 10001).isEligibleForComparison, isFalse);
    expect(offer(price: 620, offerUrl: Uri.parse('https://user:secret@example.com/offer')).isEligibleForComparison, isFalse);
    expect(offer(price: 620, offerUrl: Uri.parse('https:/offer')).isEligibleForComparison, isFalse);
    expect(offer(price: 620, offerUrl: Uri.parse('https://localhost/offer')).isEligibleForComparison, isFalse);
    expect(offer(price: 620, offerUrl: Uri.parse('https://partner.local/offer')).isEligibleForComparison, isFalse);
  });

  test('excludes non-public network offer URLs', () {
    for (final host in ['127.0.0.1', '10.0.0.8', '100.64.0.1', '100.127.255.254', '169.254.10.20', '172.16.0.1', '172.31.255.254', '192.0.0.1', '192.0.2.1', '192.168.1.20', '198.18.0.1', '198.51.100.1', '203.0.113.1', '224.0.0.1', '255.255.255.255', '[::1]', '[::]', '[fc00::1]', '[fd12:3456::1]', '[fe80::1]']) {
      expect(
        offer(price: 620, offerUrl: Uri.parse('https://$host/offer')).isEligibleForComparison,
        isFalse,
        reason: host,
      );
    }
    expect(offer(price: 620, offerUrl: Uri.parse('https://100.128.0.1/offer')).isEligibleForComparison, isTrue);
    expect(offer(price: 620, offerUrl: Uri.parse('https://172.32.0.1/offer')).isEligibleForComparison, isTrue);
    expect(offer(price: 620, offerUrl: Uri.parse('https://[2606:4700:4700::1111]/offer')).isEligibleForComparison, isTrue);
  });

  test('excludes stale offers when a comparison time is supplied', () {
    final now = DateTime.parse('2026-09-16T10:00:00Z');
    final best = bestComparableBuybackOffer(
      [offer(price: 700, checkedAt: now.subtract(const Duration(hours: 25))), offer(price: 620, checkedAt: now.subtract(const Duration(hours: 2)))],
      condition: BuybackCondition.likeNew,
      now: now,
    );
    expect(best?.price, 620);
  });

  test('tolerates small clock skew but rejects implausible future timestamps', () {
    final now = DateTime.parse('2026-09-16T10:00:00Z');
    expect(offer(price: 620, checkedAt: now.add(const Duration(minutes: 4))).isFreshAt(now), isTrue);
    expect(offer(price: 700, checkedAt: now.add(const Duration(minutes: 6))).isFreshAt(now), isFalse);
  });

  test('parses normalized adapter payload', () {
    final parsed = BuybackOffer.fromJson(payload());
    expect(parsed.condition, BuybackCondition.likeNew);
    expect(parsed.price, 620);
    expect(parsed.isEligibleForComparison, isTrue);
  });

  test('rejects non-buyback price kinds', () {
    final json = payload()..['price_kind'] = 'asking_price';
    expect(() => BuybackOffer.fromJson(json), throwsFormatException);
  });
}

File: server/buyback_test.js
'use strict';

const assert = require('assert');
const { normalizeBuybackPayload, bestBuybackOffer } = require('./buyback');

const now = Date.parse('2026-09-16T10:00:00Z');
function offer(overrides = {}) {
  return {
    provider_id: 'provider-a',
    provider_name: 'Provider A',
    product_id: 'iphone-15-pro-256',
    matched_title: 'Apple iPhone 15 Pro 256 GB',
    condition: 'like_new',
    price: 620,
    currency: 'EUR',
    offer_url: 'https://example.com/offer',
    checked_at: '2026-09-16T09:30:00Z',
    price_kind: 'indicative_buyback',
    requires_inspection: true,
    match_confidence: 0.98,
    ...overrides,
  };
}

const normalized = normalizeBuybackPayload({ items: [offer()] }, { now });
assert.strictEqual(normalized.length, 1);
assert.strictEqual(normalized[0].price, 620);

assert.strictEqual(normalizeBuybackPayload({ items: [offer({ checked_at: '2026-09-14T09:30:00Z' })] }, { now }).length, 0);
assert.strictEqual(normalizeBuybackPayload({ items: [offer({ price: 0 })] }, { now }).length, 0);
assert.strictEqual(normalizeBuybackPayload({ items: [offer({ price: 10001 })] }, { now }).length, 0, 'implausible feed prices must not distort deal ranking');
assert.strictEqual(normalizeBuybackPayload({ items: [offer({ price_kind: 'asking_price' })] }, { now }).length, 0);
assert.strictEqual(normalizeBuybackPayload({ items: [offer({ match_confidence: 0.72 })] }, { now }).length, 0);
assert.strictEqual(normalizeBuybackPayload({ items: [offer({ condition_uncertain: true })] }, { now }).length, 0);
assert.strictEqual(normalizeBuybackPayload({ items: [offer({ offer_url: 'http://example.com/offer' })] }, { now }).length, 0);
assert.strictEqual(normalizeBuybackPayload({ items: [offer({ offer_url: 'https://user:secret@example.com/offer' })] }, { now }).length, 0, 'offer URLs must not embed credentials');

const comparable = normalizeBuybackPayload({ items: [
  offer({ provider_id: 'a', price: 610 }),
  offer({ provider_id: 'b', price: 645 }),
  offer({ provider_id: 'c', price: 700, condition: 'used_good' }),
] }, { now });
assert.strictEqual(bestBuybackOffer(comparable, 'like_new').provider_id, 'b');
assert.strictEqual(bestBuybackOffer(comparable, 'like_new').price, 645);

const duplicates = normalizeBuybackPayload({ items: [
  offer({ provider_id: 'a', price: 610, checked_at: '2026-09-16T09:20:00Z' }),
  offer({ provider_id: 'a', price: 625, checked_at: '2026-09-16T09:10:00Z' }),
  offer({ provider_id: 'a', price: 625, checked_at: '2026-09-16T09:40:00Z' }),
  offer({ provider_id: 'b', price: 640 }),
] }, { now });
assert.strictEqual(duplicates.length, 2, 'same provider/product/condition must not inflate comparison');
assert.strictEqual(duplicates[0].provider_id, 'b', 'comparison should be ordered by best payout');
assert.strictEqual(duplicates[1].price, 625);
assert.strictEqual(duplicates[1].checked_at, '2026-09-16T09:40:00.000Z', 'equal-price duplicate should keep freshest quote');

console.log('buyback tests passed');
